---

<h1 align=left><font size = 8, style="color:rgb(200,0,0)"><b>Introdução  </b></font></h1>

---

**Aprendizado de máquina**

(em inglês, "machine learning") é uma subárea da _inteligência artificial_ que se concentra no desenvolvimento de algoritmos e modelos que permitem aos computadores aprender e melhorar o desempenho em tarefas específicas a partir da experiência com dados, em vez de serem explicitamente programados para realizar uma tarefa.


<figure>
<center>
<img src="https://visibilia.net.br/wp-content/uploads/2020/12/AM-2.png" width=700, align="centre"/>
<figcaption>Divisão dos dados</figcaption></center>
</figure>
<br>

[link: imagem]("https://visibilia.net.br/wp-content/uploads/2020/12/AM-2.png")


> A ideia fundamental do aprendizado de máquina é que **os sistemas computacionais podem analisar e interpretar dados, identificar padrões e fazer previsões ou tomar decisões com base nesses padrões, tudo de forma autônoma.** Isso é feito através de uma variedade de técnicas e algoritmos que permitem que as máquinas "aprendam" a partir dos dados de treinamento.

<figure>
<center>
<img src="https://github.com/Rogerio-mack/Machine-Learning-I/raw/main/Figures/MLparadigmas/Slide2.PNG" width=700, align="centre"/>
<figcaption>Divisão dos dados</figcaption></center>
</figure>
<br>

[link: imagem]("https://github.com/Rogerio-mack/Machine-Learning-I/raw/main/Figures/MLparadigmas/Slide2.PNG")

# <h align=left><font size = 6, > Cases: Aprendizado de Máquina</font></h>

Ao longo do texto demos vários exemplos de aplicações que vem sendo feitas do Aprendizado de Máquina. Para algumas empresas o Aprendizado de Máquina está no centro de seus negócios. É caso das redes sociais, do Facebook e o Instagram, ao LinkedIn, ou a Netflix e o sistema de busca da Google. Mas de empresas do varejo à empresas de extração, saúde ou financeiras, todos os setores e mesmo a Ciência, estão adotando o Aprendizado de Máquina, mesmo que não sendo esse o seu propósito central. Revisamos aqui alguns dos usos mais comuns que encontram-se espalhados ao longo do texto para ilustrar aspectos específicos do Aprendizado de Máquina.

## Sistemas de recomendação

Recomendações da Netflix, YouTube ou Instagram, de mensagens do Twitter ou de anúncios no FaceBook que queremos ver buscam extrair o padrão de preferência dos usuários.

<figure>
<center>
<img src="https://slideplayer.com.br/slide/1594488/5/images/15/Sistemas+de+recomenda%C3%A7%C3%A3o.jpg" width=700, align="centre"/>
<figcaption>Divisão dos dados</figcaption></center>
</figure>
<br>

[link: imagem]("https://slideplayer.com.br/slide/1594488/5/images/15/Sistemas+de+recomenda%C3%A7%C3%A3o.jpg")


## Detecção de fraude

>> Identificação de transações de cartão de crédito ou compras fraudulentas em sites de e-commerce, tentativas fraudulentas de login etc.

## Chatbots

>> Chatbots online em que os clientes, interagem por texto ou mesmo fala, com uma máquina que guia o usuário para um determinado resultado empregam uma série de recursos de Aprendizado de Máquina, do processamento de linguagem natural ao aprendizado de reforço.

## Diagnóstico médico

>> Algoritmos de aprendizado de máquina são empregados para examinar imagens médicas ou outros dados de pacientes e auxiliar no diagnóstico precoce de doenças como câncer ou diabetes.

## Veículos Autônomos


<img src="http://revistapesquisa.fapesp.br/wp-content/uploads/2015/09/Veiculos_00057830-300x212.jpg" width=500, align="center">

<small>Fonte: https://revistapesquisa.fapesp.br/o-futuro-sem-motorista/



>> Empregam um grande número de recursos de aprendizado de máquina para segmentação de cenas (identificação de ruas, calçadas, prédios e pessoas em uma cena), movimentação autônoma, predição e tratamento de colisão etc. em geral com emprego de algoritmos de Deep Learning.


<img src="https://revistapesquisa.fapesp.br/wp-content/uploads/2019/12/068-073_evtol_286-1-IMG.png" width=500, align="center">

<small><b>
eVTOL aeronave elétrica de decolagem e pouso vertical sendo desenvolvido pela EMBRAER-UBER. Fonte: https://revistapesquisa.fapesp.br/2019/12/03/a-chegada-dos-carros-voadores/

## Análise e Classificação de Risco

>> Para precificação de crédito e seguros.


<img src="https://revistapesquisa.fapesp.br/wp-content/uploads/2017/07/068-071_madeira_257-3.jpg" width=500, align="center">

<small><b>
Figura 8. Sistema Automático de identificação de lâminas de madeira para uso industrial. Fonte: https://revistapesquisa.fapesp.br/2017/07/18/identificacao-de-madeiras/


## Exploração eficaz do Presál - PETROBRÁS


<img src="https://revistapesquisa.fapesp.br/wp-content/uploads/2021/04/SITE_Petroleo-0-1140.jpg" width=500, align="center">

<small><b>
Figura 9. Sistema Híbrido combina Mecânica dos Fluídos e Aprendizado de Máquina para exploração mais eficaz do Pré-Sal pela Petrobrás. Fonte: https://revistapesquisa.fapesp.br/producao-mais-eficaz-no-pre-sal/


## Tratamento de agua


<img src="https://revistapesquisa.fapesp.br/wp-content/uploads/2020/03/070-074_mineracao-digital_289-3-1140.jpg" width=500, align="center">

<small><b>
Figura 10. Mina de alumínio usa sistema automatizado para tratar água de rejeito. Fonte: https://revistapesquisa.fapesp.br/mineracao-digital/


## Outras aplicações

<br>
<br>

<img src="https://github.com/firmai/industry-machine-learning/raw/master/assets/industry.png" width=500, align="center">

[Machine Learning and Data Science Applications in Industry](https://github.com/ashishpatel26/Real-time-ML-Project) | https://github.com/ashishpatel26/Real-time-ML-Project

| <!-- -->                         | <!-- -->                         | <!-- -->                          |
| -------------------------------- | -------------------------------- | --------------------------------- |
| [Accommodation & Food](#accommodation)             | [Agriculture](#agriculture)           | [Banking & Insurance](#bankfin)               |
| [Biotechnological & Life Sciences](#biotech) | [Construction & Engineering](#construction)       | [Education & Research](#education)              |
| [Emergency & Relief](#emergency)               | [Finance](#finance) | [Manufacturing](#manufacturing)             |
| [Government and Public Works](#public)      | [Healthcare](#healthcare)  | [Media & Publishing](#media)                |
| [Justice, Law and Regulations](#legal)      | [Miscellaneous](#miscellaneous)                    | [Accounting](#accounting) |
| [Real Estate, Rental & Leasing](#realestate)    | [Utilities](#utilities)              | [Wholesale & Retail](#wholesale)                  |

<br>
<br>









# EXEMPLO


Em seguida, iremos responder de duas formas diferentes á pergunta: Um grau celsiu a quantos Fahrenheit equivale?

1.   Utilizando equações
2.   Reconhecimento de padrões

> Você certamente não terá dificuldade de criar um programa para converter temperaturas de $^{o} C$ (Celsius) para $^{o} F$ (Fahrenheit).

<img src="https://github.com/Rogerio-mack/Machine-Learning-I/raw/main/Figures/MLparadigmas/Slide1.PNG" width=800, align="center">


## 1. Equações matematicas

> Lembre-se daquel assunto que vimos em Analise exploratorio de dados chamado de "funções "!! então aqui vamos construir uma dessas funções

In [ ]:
def Celsius_para_Fahrenheit(Celsius):

  return Celsius * 9/5 + 32

>> Vamos testar, troque o "X" pelos último digito da sua matricula

In [ ]:
print( Celsius_para_Fahrenheit( 9  ) )

48.2


>> Vamos testar novamente, troque o "Y" pelo primeiro digito da sua matricula


In [ ]:
print( Celsius_para_Fahrenheit( 9 ) )

48.2


## 2. Reconhecimento de padrões: Aprendizado de maquina

>> Mas será que um **'programa'** pode criar esse programa de conversão de temperaturas? A ideia do Aprendizado de Máquina é que sim. Mas para que seja possível o aprendizado precisamos partir de algum ponto que são os dados.

### Importamos as bibliotecas que iremos utilizar

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns

### Geramos/importamos os dados

In [ ]:
# Criamos uma base de dados vazia para comportar os dados que em seguida vamos gerar
df = pd.DataFrame()

# Criando um arranjo de dados para iniciar com o aprendizado de máquina
df['celsius'] = np.arange(-20,60)

# Após criar o arranjo utilizamos o mini-programa que criamos acima para criar uma nova coluna chamada "Fahrenheit" e convertimos os valores da coluna de Celsius para Fahrenheit
df['fahrenheit'] = Celsius_para_Fahrenheit( df['celsius'] )

# Agora vamos lêr as primeras linhas do nosso arranjo
df.head(40)


celsius  fahrenheit
0       -20        -4.0
1       -19        -2.2
2       -18        -0.4
3       -17         1.4
4       -16         3.2
5       -15         5.0
6       -14         6.8
7       -13         8.6
8       -12        10.4
9       -11        12.2
10      -10        14.0
11       -9        15.8
12       -8        17.6
13       -7        19.4
14       -6        21.2
15       -5        23.0
16       -4        24.8
17       -3        26.6
18       -2        28.4
19       -1        30.2
20        0        32.0
21        1        33.8
22        2        35.6
23        3        37.4
24        4        39.2
25        5        41.0
26        6        42.8
27        7        44.6
28        8        46.4
29        9        48.2
30       10        50.0
31       11        51.8
32       12        53.6
33       13        55.4
34       14        57.2
35       15        59.0
36       16        60.8
37       17        62.6
38       18        64.4
39       19        66.2

>> Se tivermos um conjunto de dados com valores de temperaturas em $^{o} C$ (Celsius) e seus respectivos valores em $^{o} F$ (Fahrenheit), podemos criar um programa que *aprende* esse padrão e empregá-lo para a conversão de novos dados.

### Importamos as funções das bibliotecas para treinar o nosso modelo de Regressão

In [ ]:
# vamos iniciar importanto a libraria "sklearn" e utilizaremos a regressão linear para que o nosso modelo consiga aprender
from sklearn.linear_model import LinearRegression

### Definimos as nossas variveis preditoras

In [ ]:
# Variaveis de entrada
X = df[['celsius']]

# Variaveis de saida
y = df['fahrenheit']

### Inicializamos o nosso modelo de aprendizagem [REGRESSÃO]

In [ ]:
regressor = LinearRegression()
regressor.fit(X, y)

LinearRegression()

>> O programa acima cria um *modelo* (um modelo linear) para os dados. Esse modelo busca capturar o *padrão* dos dados, isto é, a regra que leva os valores de $^{o} C$ (Celsius) $\rightarrow$ $^{o} F$ (Fahrenheit), e cria um *programa* com essa regra que pode ser aplicado a novos valores.

- Em seguida, obtemos a predição do valor que desejamos, nesse caso será o mesmo que utilizamos na primeira parte para realizar a comparação

In [ ]:
Fahrenheit_prediction = regressor.predict(pd.DataFrame({'celsius':[0,27]}))
print(Fahrenheit_prediction)

[32.  80.6]


>> Agora, nosso novo programa `regressor`, converte as temperaturas da mesma forma que o programa `Celsius2Fahrenheit` onde explicitamente codificamos a regra de conversão de Celsius para Fahrenheit, mas *criamos* esse programa unicamente a partir de dados, sem conhecermos explicitamente a regra de conversão.

### Coeficientes do modelo de Regressão

In [ ]:
print("Coefficients: \n", regressor.coef_, regressor.intercept_)

Coefficients: 
 [1.8] 32.0


> Nosso `regressor` é um modelo de regressão linear, um dos modelos de Aprendizado de Máquina mais simples, e que mapeia dados de entrada e saída na forma de uma equação linear.

$$ y = a_0 + a_1 x $$

O *aprendizado* consiste em encontrar, nesse modelo, os coeficientes que melhor se ajustam aos dados e você pode notar que os coeficientes obtidos são exatamente os mesmos da nossa fórmula colocada explicitamente no nosso código `Celsius_para_Fahrenheit`.

$$ y = 32 + 1.8 x \space \Rightarrow \space ^{o}C = 32 + \frac{5}{9} \times ^{o}F $$

>> Talvez você ache pouco interessante o exemplo que apresentamos. Afinal parece ter sido muito mais fácil criar o programa `Celsius_para_Fahrenheit` e, de fato, quando conhecemos a regra por traz dessas transformações é muito mais simples explicitarmos isso em um programa que empregarmos o Aprendizado de Máquina. Um modelo conhecido (a regra explícita) apresentará sempre um resultado melhor.

<br>**Mas e quando você não conhece essa regra?**

>> Há muitos problemas práticos em que não conhecemos essa regra (ela pode até mesmo não existir) e não podemos, portanto, explicitá-la em um programa. É nesses casos que o Aprendizado de Máquina será útil.

>